### PR TITLE
[Mono] Enable AOT to print out the list of compiled methods

### LINKED ISSUE #79551: [Mono] Enable Mono AOT compiler to print out the list of compiled methods
This feature will help with IL stripping for MAUI Android apps. (https://github.com/dotnet/runtime/issues/44855)
The list of compiled methods will be passed into the IL stripping tool to remove AOT compiled methods.

Output format: text file

### PR DESCRIPTION
**Pull request by fanyang-mono**
Print out tokens of compiled methods into one text file per assembly. To enable this feature, the user needs to provide a location to save these files, by adding aot switch `compiled-methods-outpath=<desired_location>`. For each aot compiled assembly, there will be a corresponding text file being generated into that location. For example, the compile method token list of `System.Private.CoreLib.dll` will be saved into `System_Private_CoreLib_dll_compiled_methods.txt`.

The compiled method tokens will later be used by an IL trimming tool to trim away compiled method IL code. That way it makes Maui Android app smaller and app creator's IP will be protected as well.

Fixes #79551

### CODE REVIEW

File: src/mono/mono/mini/aot-compiler.c [hunk 1]
```diff
@@ -14142,6 +14151,14 @@ aot_assembly (MonoAssembly *ass, guint32 jit_opts, MonoAotOptions *aot_options)
 		acfg->logfile = fopen (acfg->aot_opts.logfile, "a+");
 	}
 
+	if (acfg->aot_opts.compiled_methods_outpath) {
```
Comment (vargaz) [inline src/mono/mono/mini/aot-compiler.c:14154]: Wouldn't be better for the caller of the aot compiler passing in the full path ? It has to compute it anyway if it wants to read it afterwards.

File: src/mono/mono/mini/aot-compiler.c [hunk 2]
```diff
@@ -14455,6 +14472,10 @@ aot_assembly (MonoAssembly *ass, guint32 jit_opts, MonoAotOptions *aot_options)
 
 	current_acfg = NULL;
 
+	if (acfg->aot_opts.compiled_methods_outpath) {
+		fclose (acfg->compiled_methods_outfile);
```
Comment (kotlarmilos) [inline src/mono/mono/mini/aot-compiler.c:14476]: If dedup is enabled, `aot_assembly` could return before `fclose` and leave file descriptor dangling. Not sure if it is the only case, there might be more.

File: src/mono/mono/mini/aot-compiler.c (hunk 1 above)
Comment (fanyang-mono) [reply src/mono/mono/mini/aot-compiler.c:14154]: Fixed

File: src/mono/mono/mini/aot-compiler.c (hunk 2 above)
Comment (fanyang-mono) [reply src/mono/mono/mini/aot-compiler.c:14476]: That's a good point. I added the check when open the file. if `acfg->dedup_collect_only` is true, aot won't print out compiled method list.

**Review by vargaz — APPROVED** [review]

**Comment by fanyang-mono** [comment]
CI failure is a known issue #81123

**Pull request merged**